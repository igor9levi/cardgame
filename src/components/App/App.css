.App {
  text-align: center;
  background-color: darkgreen;
  margin: 30px auto auto;
  display: flex;
  flex-direction: column;
  flex: 1;
  align-items: center;
  justify-content: center;
}

