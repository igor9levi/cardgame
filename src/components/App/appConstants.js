
export const NUM_ROUNDS = 10;
export const START_GAME = 'START_GAME';

export const SET_DECK_ID = 'SET_DECK_ID';

export const INITIAL_STATE = {
  numPlayers: 0,
  deckId: null,
  players: [],
};
