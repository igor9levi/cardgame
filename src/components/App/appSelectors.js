export const getNumPlayers = state => state.app.numPlayers;
