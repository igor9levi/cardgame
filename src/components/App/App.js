import React, { Component } from 'react';
import './App.css';
import Board from '../Board';
import Welcome from '../Welcome/';

export class App extends Component {
  render() {
    return (
      <div className="App">
        <Welcome />
      </div>
    );
  }
}

export default App;
