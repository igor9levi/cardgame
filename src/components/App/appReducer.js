import * as constants from './appConstants';

export default (state = constants.INITIAL_STATE, action) => {
  switch (action.type) {
    case constants.START_GAME:
      return {
        ...state,
        numPlayers: action.payload,
        status: constants.APP_STATUS.PLAY,
      };
    case constants.SET_DECK_ID:
      return {
        ...state,
        deckId: action.payload,
      };
    default:
      return state;
  }
};
