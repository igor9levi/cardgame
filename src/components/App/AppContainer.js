import { connect } from 'react-redux';
import AppComponent from './App';
import { getAppStatus } from './appSelectors';

export const mapStateToProps = state => ({
  status: getAppStatus(state),
});

export default connect(mapStateToProps, null)(AppComponent);
