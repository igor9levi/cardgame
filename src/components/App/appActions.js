import * as playerConstants from '../../modules/Player/playerConstants';
import * as appConstants from './appConstants';
import * as selectors from './appSelectors';
import apiProvider from '../../modules/api/ApiProvider';

export const setWelcomeStatus = () => ({
  type: appConstants.GAME_RESET,
});

export const setPlayStatus = () => ({
  type: appConstants.GAME_READY,
});

export const setEndStatus = () => ({
  type: appConstants.GAME_END,
});

export const setNumPlayers = numPlayers => ({
  type: appConstants.GAME_START,
  payload: numPlayers,
});

export const setError = error => ({
  type: appConstants.SET_ERROR,
  payload: error,
});

export const setDeck = ({ deckId }) => ({
  type: appConstants.SET_DECK_ID,
  payload: deckId,
});

export const setCards = ({ cards }) => ({
  type: playerConstants.SET_CARDS,
  payload: cards,
});

export const getDeck = () => apiProvider.getDeck();

export const getCards = async ({ deckId, numPlayers }) => {
  const cards = [];
  const response = await apiProvider.getCards({ deckId, numCards: appConstants.NUM_ROUNDS * numPlayers });
  for (let i = 0; i < numPlayers; i++) {
    const cardsForOnePlayer = response.slice(i * appConstants.NUM_ROUNDS, ((i + 1) * appConstants.NUM_ROUNDS));
    const cardsWithPlayerId = cardsForOnePlayer.map(card => ({ ...card, playerId: i }));
    cards.push(cardsWithPlayerId);
  }

  return cards;
};

export const startGame = numPlayers => async (dispatch, getState) => {
  dispatch(setNumPlayers(numPlayers));
  try {
    const deckId = await getDeck();
    dispatch(setDeck({ deckId }));

    const cards = await getCards({ deckId, numPlayers: selectors.getNumPlayers(getState()) });
    dispatch(setCards({ cards }));
    dispatch(setPlayStatus());
  } catch (error) {
    dispatch(setError(error));
  }
};
