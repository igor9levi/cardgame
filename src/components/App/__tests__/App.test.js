import React from 'react';
import { shallow } from 'enzyme';
import App from '../App';
import { APP_STATUS } from '../appConstants';

describe('Test App', () => {
  it('renders App without crashing status welcome', () => {
    const wrapper = shallow(<App status={APP_STATUS.WELCOME} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders App without crashing status play', () => {
    const wrapper = shallow(<App status={APP_STATUS.PLAY} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders App without crashing status end', () => {
    const wrapper = shallow(<App status={APP_STATUS.END} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders App without crashing status loading', () => {
    const wrapper = shallow(<App status={APP_STATUS.LOADING} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders App without crashing status error', () => {
    const wrapper = shallow(<App status={APP_STATUS.ERROR} />);
    expect(wrapper).toMatchSnapshot();
  });
});
