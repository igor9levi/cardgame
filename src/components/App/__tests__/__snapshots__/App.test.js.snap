// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test App renders App without crashing status end 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App
    status={4}
    store={
        Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Connect(ScoreBoard) />,
      "className": "App",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Connect(ScoreBoard) />,
        "className": "App",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`Test App renders App without crashing status error 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App
    status={5}
    store={
        Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Connect(Component) />,
      "className": "App",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Connect(Component) />,
        "className": "App",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`Test App renders App without crashing status loading 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App
    status={2}
    store={
        Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <h1>
        Loading...
</h1>,
      "className": "App",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Loading...",
      },
      "ref": null,
      "rendered": "Loading...",
      "type": "h1",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <h1>
          Loading...
</h1>,
        "className": "App",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Loading...",
        },
        "ref": null,
        "rendered": "Loading...",
        "type": "h1",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`Test App renders App without crashing status play 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App
    status={3}
    store={
        Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Connect(Board) />,
      "className": "App",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Connect(Board) />,
        "className": "App",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`Test App renders App without crashing status welcome 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App
    status={1}
    store={
        Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Connect(Component) />,
      "className": "App",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Connect(Component) />,
        "className": "App",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
